enhancements/microshift/microshift-greenboot.md

```diff
@@ -0,0 +1,194 @@
+be installed into `/etc/greenboot/check/required.d` directory conditional to
+the presence of the `greenboot` package. This would prevent unnecessary scripts
+from being installed on systems that are not `rpm-ostree` based.
+
+The health check script will implement the following validations for the core
+MicroShift namespaces:
+- openshift-ovn-kubernetes
+- openshift-service-ca
+- openshift-ingress
+- openshift-dns
+- openshift-storage
+
+> When pluggable CNI and CSI functionality is implemented, the `openshift-ovn-kubernetes`
+> and `openshift-storage` namespace name selection need to be conditional on
+> the loaded CNI and CSI plugins.
+
+Exiting the script with a non-zero status will have the boot declared as failed.
+
+|Validation                                        |Pass  |Fail  |
+|--------------------------------------------------|------|------|
+|microshift.service is enabled                     |Next  |exit 0|
+|Wait for any pods to start                        |Next  |exit 1|
+|For each core namespace, wait for pods to be ready|Next  |exit 1|
+|For each core namespace, check pods not restarting|exit 0|exit 1|
+
+The wait period in each validation starts from 5 minutes and it is doubled
+after each boot in the verification loop. It should be possible to override the
+base wait period setting.
+
+### Risks and Mitigations
+
+Edge Devices often have relatively slow networks or occasional loss of network
+connectivity. The risk of declaring a `false positive` on the failed boot may
+be caused by these factors.
+
+If an update requires downloading new container images on a relatively slow
+network, it may take more than the default wait period to run this operation.
+The mitigation would be automatic doubling of the wait period on subsequent
+boots. If this is still not enough, system administrators should set the base
+wait period setting to a higher value to start with.
+
+If any external connectivity is required for upgrades to complete successfully,
+intermittent network outage may result in a rollback to the previous version.
+As a mitigation, system administrators should reinstall the upgrade when network
+is available for the Edge Devices.
+
+To further mitigate potential problems related to slow or unstable networks,
+system administrators may choose to bundle the upgraded container images in the
+`ostree` software upgrade. In this case, when the upgrade is installed, no
+more container images will need to be downloaded over the network.
+
+### Drawbacks
+
+The technical implementation of upgrade validation is rather simple as it relies
+on the existing `greenboot` and `rpm-ostree` facilities. However, end users may
+find it annoying to have their Edge Devices rebooted few times until the upgrade
+is either successful or rolled back.
+
+It is possible that certain 3rd party workloads become operational before the
+upgrade is declared successful. This situation may further hurt user experience
```

Comment:
@romfreiman, can you explain what you mean here?

Reply:
as you wrote, the 3rd party workload can start, while microshift is not ready yet, and it can write down some data that is related to the upgrade. And then u decide to roll back.....

Reply:
I added this case in Risks and Mitigations section.